src/streaming/array.writer.cpp

```diff
@@ -0,0 +1,422 @@
+#include "macros.hh"
+#include "array.writer.hh"
+#include "zarr.common.hh"
+#include "zarr.stream.hh"
+#include "sink.creator.hh"
+
+#include <cmath>
+#include <functional>
+#include <latch>
+#include <stdexcept>
+
+#ifdef min
+#undef min
+#endif
+
+bool
+zarr::downsample(const ArrayWriterConfig& config,
+                 ArrayWriterConfig& downsampled_config)
+{
+    // downsample dimensions
+    std::vector<ZarrDimension> downsampled_dims;
```

Comment:
If you know the size of `downsampled_dims` (I assume it's `config.dimensions->ndims`), you should use `std::array`. If you must use a dynamically sized array, you can allocate space for it ahead of time to avoid unnecessary allocations.